### PR TITLE
Fix UTF-16 TextDecoder unpaired surrogate handling (#4612)

### LINKED ISSUE #4612: utf-16 TextDecoder is incorrect
It should always produce well-formed strings

Instead, this is what happens on `utf-16le` and `utf-16be`:
```
✖ FAIL Common implementation mistakes > Unicode > Invalid input is replaced > utf-16le
  AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
+ actual - expected

+ 'ab\ud800wx'
- 'ab�wx'
     ^
✖ FAIL Common implementation mistakes > Unicode > Invalid input is replaced > utf-16be
  AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
+ actual - expected

+ 'ab\ud800wx'
- 'ab�wx'
     ^
```

Libraries/users might rightfully assume that TextDecoder always produces well-formed strings, and not doing so is problematic.

This might have security implications due to:
1. how hashes and signatures behave on non-well-formed strings
2. not following `fatal` option

### PR DESCRIPTION
**Pull request by Rahul-2k4**
## Summary
- make UTF-16 TextDecoder output well-formed strings by replacing unpaired surrogates with `U+FFFD`
- apply the fix to both `utf-16le` and `utf-16be` decode paths
- keep truncation behavior correct for odd-length input (no extra replacement when the last decoded code unit is a high surrogate)

## Tests
- add regression tests for unpaired surrogate replacement in `utf-16le` and `utf-16be`
- add regression tests for truncation edge cases in both endian modes
- verified with:
  - `cargo test -p boa_runtime decoder_utf16`
  - `cargo test -p boa_runtime`
  - `cargo fmt --all --check`
  - `cargo clippy -p boa_runtime --all-targets -- -D warnings`

Closes #4612.

### CODE REVIEW

File: core/runtime/src/text/encodings.rs [hunk 1]
```diff
@@ -59,28 +69,20 @@ pub(crate) mod utf16be {
...
-
-        if dangling {
-            JsString::from(&[JsString::from(&*input), js_string!("\u{FFFD}")])
-        } else {
-            JsString::from(&*input)
-        }
+        let code_units = input
+            .chunks_exact(2)
+            .map(|pair| u16::from_be_bytes([pair[0], pair[1]]))
+            .collect::<Vec<_>>();
```
Comment (jedel1043) [inline core/runtime/src/text/encodings.rs]: Why collect here if you can just pass the iterator? We're already using an iterator in `decode_utf16_units`.

**Comment by zhuzhu81998** [comment]
@Rahul-2k4 are you still working on this?

**Comment by Rahul-2k4** [comment]
yep just finished in a while rn final commits are taking place

File: core/runtime/src/text/encodings.rs (hunk 1 above)
Comment (Rahul-2k4) [reply core/runtime/src/text/encodings.rs]: yep its done now

File: core/runtime/src/text/tests.rs [hunk 2]
```diff
@@ -125,6 +125,110 @@ fn decoder_js_invalid() {
...
+                const d = new TextDecoder("utf-16le");
+                for (const { invalid, replaced } of invalid16) {
+                    const input = new Uint8Array(invalid.length * 2);
+                    for (let i = 0; i < invalid.length; i++) {
+                        input[2 * i] = invalid[i] & 0xff;
+                        input[2 * i + 1] = invalid[i] >> 8;
+                    }
+                    assertEq(d.decode(input), String.fromCharCode(...replaced));
+                }
+            "#}),
```
Comment (zhuzhu81998) [inline core/runtime/src/text/tests.rs]: in the spirit of unit tests, it would probably be better to not do this part as JS code. For example if `String.fromCharcode` has a problem this would fail even though `decode` works fine.

File: core/runtime/src/text/tests.rs [hunk 3]
```diff
@@ -125,6 +125,110 @@ fn decoder_js_invalid() {
...
+                ];
+
+                const d = new TextDecoder("utf-16be");
+                for (const { invalid, replaced } of invalid16) {
+                    const input = new Uint8Array(invalid.length * 2);
+                    for (let i = 0; i < invalid.length; i++) {
+                        input[2 * i] = invalid[i] >> 8;
+                        input[2 * i + 1] = invalid[i] & 0xff;
+                    }
+                    assertEq(d.decode(input), String.fromCharCode(...replaced));
```
Comment (zhuzhu81998) [inline core/runtime/src/text/tests.rs]: same here

**Comment by zhuzhu81998** [comment]
Hmm why push commits after approval and then close it before it could be merged?
